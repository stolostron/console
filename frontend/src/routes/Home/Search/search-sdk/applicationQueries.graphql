query getApplications {
    applications {
        _uid
        apiVersion
        created
        dashboard
        labels
        name
        namespace
        cluster
        clusterCount
        hubChannels
        hubSubscriptions {
            _uid
            name
            localPlacement
            timeWindow
            status
            channel
        }
        applicationSet
        destinationName
        destinationServer
        destinationCluster
        destinationNamespace
        repoURL
        path
        chart
        targetRevision
    }
}

query getSubscriptions {
    subscriptions {
        _uid
        name
        namespace
        created
        timeWindow
        localPlacement
        status
        channel
        appCount
        clusterCount
    }
}

query getPlacementRules {
    placementRules {
        _uid
        name
        namespace
        created
        replicas
        clusterCount
    }
}

query getChannels {
    channels {
        _uid
        name
        namespace
        created
        type
        pathname
        localPlacement
        subscriptionCount
        clusterCount
    }
}
