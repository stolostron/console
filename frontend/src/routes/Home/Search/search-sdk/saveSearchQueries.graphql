mutation saveSearch($resource: JSON!) {
    saveSearch(resource: $resource)
}

mutation deleteSearch($resource: JSON!) {
    deleteSearch(resource: $resource)
}
