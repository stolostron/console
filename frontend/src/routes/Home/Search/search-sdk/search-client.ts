/* Copyright Contributors to the Open Cluster Management project */
// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project

import { ApolloClient, ApolloLink, from, HttpLink, InMemoryCache } from '@apollo/client'
import { getBackendUrl, getCookie } from '../../../../resources'

const httpLink = new HttpLink({
  uri: () => `${getBackendUrl()}/proxy/search`,
})

const csrfHeaderLink = new ApolloLink((operation, forward) => {
  const csrfToken = getCookie('csrf-token')
  if (csrfToken) {
    operation.setContext(({ headers = {} }) => ({
      headers: {
        ...headers,
        'X-CSRFToken': csrfToken,
      },
    }))
  }

  return forward(operation)
})

export const searchClient = new ApolloClient({
  connectToDevTools: process.env.NODE_ENV === 'development',
  link: from([csrfHeaderLink, httpLink]),
  cache: new InMemoryCache(),
  credentials: 'same-origin',

  defaultOptions: {
    watchQuery: {
      fetchPolicy: 'network-only',
      errorPolicy: 'all',
    },
    query: {
      fetchPolicy: 'network-only',
      errorPolicy: 'all',
    },
    mutate: {
      errorPolicy: 'all',
    },
  },
})
