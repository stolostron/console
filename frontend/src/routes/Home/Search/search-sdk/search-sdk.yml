# Copyright (c) 2021 Red Hat, Inc.
# Copyright Contributors to the Open Cluster Management project
overwrite: true
schema: https://localhost:4010/searchapi/graphql
documents: ./src/search-sdk/*.graphql
generates:
  src/search-sdk/search-sdk.ts:
    plugins:
      - typescript
      - typescript-operations
      - typescript-react-apollo
    config:
      skipTypename: true
