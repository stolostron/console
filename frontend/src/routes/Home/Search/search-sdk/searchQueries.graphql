query searchSchema {
    searchSchema
}

query searchComplete($property: String!, $query: SearchInput, $limit: Int) {
    searchComplete(property: $property, query: $query, limit: $limit)
}

query searchResultItems($input: [SearchInput]) {
    searchResult: search(input: $input) {
        items
    }
}

query searchResultCount($input: [SearchInput]) {
    searchResult: search(input: $input) {
        count
    }
}

query searchResultCountAndRelatedCount($input: [SearchInput]) {
    searchResult: search(input: $input) {
        count
        related {
            kind
            count
        }
    }
}

query searchResultRelatedCount($input: [SearchInput]) {
    searchResult: search(input: $input) {
        related {
            kind
            count
        }
    }
}

query searchResultRelatedItems($input: [SearchInput]) {
    searchResult: search(input: $input) {
        related {
            kind
            items
        }
    }
}

query getMessages {
    messages {
        id
        kind
        description
    }
}
