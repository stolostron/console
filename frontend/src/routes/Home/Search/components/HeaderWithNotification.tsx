/* Copyright Contributors to the Open Cluster Management project */
// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
import { Card, CardBody } from '@patternfly/react-core'
import { useTranslation } from '../../../../lib/acm-i18next'
import { NavigationPath } from '../../../../NavigationPath'
import { AcmInlineStatus, AcmPageHeader, StatusType } from '../../../../ui-components'
import { Message } from '../search-sdk/search-sdk'

export default function HeaderWithNotification(props: { messages: Message[] }) {
  const { t } = useTranslation()
  const { messages } = props

  return (
    <div style={{ outline: 'none', display: 'flex', justifyContent: 'flex-end' }}>
      <div style={{ flex: 1 }}>
        <AcmPageHeader title={t('Search')} />
      </div>

      {messages.map((msg, index) => {
        const displayShortText = t('Search is disabled on some clusters.') || msg?.description
        const displayLongText =
          t(
            'Currently, search is disabled on some of your managed clusters. Some data might be missing from the console view.'
          ) || msg?.description
        const footerText = t('View clusters with search add-on disabled.')

        return (
          <Card key={msg.id + index} style={{ border: 'none', boxShadow: 'none' }}>
            <CardBody>
              <AcmInlineStatus
                type={StatusType.warning}
                status={displayShortText}
                popover={{
                  headerContent: displayShortText,
                  bodyContent: displayLongText,
                  footerContent: msg.id === 'S20' && (
                    <a
                      href={`${NavigationPath.search}?filters={"textsearch":"kind%3ACluster%20addon%3Asearch-collector%3Dfalse%20name%3A!local-cluster"}`}
                    >
                      {footerText}
                    </a>
                  ),
                }}
              />
            </CardBody>
          </Card>
        )
      })}
    </div>
  )
}
