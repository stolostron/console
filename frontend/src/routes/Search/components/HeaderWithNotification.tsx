/* Copyright Contributors to the Open Cluster Management project */
// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
import { Card, CardBody } from '@patternfly/react-core'
import { useTranslation } from '../../../lib/acm-i18next'
import { NavigationPath } from '../../../NavigationPath'
import { useRecoilValue, useSharedAtoms } from '../../../shared-recoil'
import { AcmInlineStatus, AcmPageHeader, StatusType } from '../../../ui-components'
import { Message } from '../search-sdk/search-sdk'

export default function HeaderWithNotification(props: { messages: Message[] }) {
  const { t } = useTranslation()
  const { isGlobalHubState, settingsState } = useSharedAtoms()
  const isGlobalHub = useRecoilValue(isGlobalHubState)
  const settings = useRecoilValue(settingsState)
  const { messages } = props

  return (
    <div style={{ outline: 'none', display: 'flex', justifyContent: 'flex-end' }}>
      <div style={{ flex: 1 }}>
        <AcmPageHeader
          title={isGlobalHub && settings.globalSearchFeatureFlag === 'enabled' ? t('Global search') : t('Search')}
          titleTooltip={
            isGlobalHub &&
            settings.globalSearchFeatureFlag === 'enabled' &&
            t('Global search is enabled. Resources across all your managed hubs and clusters will be shown.')
          }
        />
      </div>

      {messages.map((msg, index) => {
        const displayShortText = t('Search is disabled on some clusters.') || msg?.description
        const displayLongText =
          t(
            'Currently, search is disabled on some of your managed clusters. Some data might be missing from the console view.'
          ) || msg?.description
        const footerText = t('View clusters with search add-on disabled.')

        return (
          <Card key={msg.id + index} style={{ border: 'none', boxShadow: 'none' }}>
            <CardBody>
              <AcmInlineStatus
                type={StatusType.warning}
                status={displayShortText}
                popover={{
                  headerContent: displayShortText,
                  bodyContent: displayLongText,
                  footerContent: msg.id === 'S20' && (
                    <a
                      href={`${NavigationPath.search}?filters={"textsearch":"kind%3ACluster%20addon%3Asearch-collector%3Dfalse%20name%3A!local-cluster"}`}
                    >
                      {footerText}
                    </a>
                  ),
                }}
              />
            </CardBody>
          </Card>
        )
      })}
    </div>
  )
}
