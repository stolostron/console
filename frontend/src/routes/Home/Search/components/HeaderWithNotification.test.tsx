/* Copyright Contributors to the Open Cluster Management project */
// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
import { render } from '@testing-library/react'
import { Message } from '../search-sdk/search-sdk'
import HeaderWithNotification from './HeaderWithNotification'

// case where we have a message about disabled search (current message)
test('renders clusters disabled message', () => {
  const disableSearch = [
    { id: 'S20', kind: 'info', description: 'Search is disabled on some of your managed clusters.' },
  ]
  const { baseElement } = render(<HeaderWithNotification messages={disableSearch} />)
  expect(baseElement).toMatchSnapshot()
})

// case where we have no message
test('renders empty message', () => {
  const emptyMessage: Message[] = []
  const { baseElement } = render(<HeaderWithNotification messages={emptyMessage} />)
  expect(baseElement).toMatchSnapshot()
})

// case where we have a different message
test('renders unknown message', () => {
  const newMessage = [{ id: 'S90', kind: 'warning', message: 'This is a new message' }]
  const { baseElement } = render(<HeaderWithNotification messages={newMessage} />)
  expect(baseElement).toMatchSnapshot()
})
