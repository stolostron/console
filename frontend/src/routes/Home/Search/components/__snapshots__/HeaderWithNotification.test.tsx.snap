// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders clusters disabled message 1`] = `
<body>
  <div>
    <div
      style="outline: none; display: flex; justify-content: flex-end;"
    >
      <div
        style="flex: 1;"
      >
        <section
          class="pf-c-page__main-section pf-m-no-padding pf-m-light"
        >
          <div
            class="pf-l-split"
          >
            <div
              class="pf-l-split__item pf-m-fill"
            >
              <div
                class="pf-l-stack pf-m-gutter"
              >
                <div
                  class="pf-l-stack__item pf-m-fill"
                >
                  <section
                    class="pf-c-page__main-section pf-m-light"
                  >
                    <div
                      class="pf-l-stack pf-m-gutter"
                    >
                      <div
                        class="pf-l-stack__item pf-m-fill"
                      >
                        <div
                          class="pf-l-split pf-m-gutter"
                        >
                          <div
                            class="pf-l-split__item"
                          >
                            <div
                              class="pf-c-content"
                            >
                              <h1
                                class="pf-c-title pf-m-2xl"
                                data-ouia-component-id="OUIA-Generated-Title-1"
                                data-ouia-component-type="PF4/Title"
                                data-ouia-safe="true"
                              >
                                Search
                              </h1>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </section>
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
      <article
        class="pf-c-card"
        data-ouia-component-id="OUIA-Generated-Card-1"
        data-ouia-component-type="PF4/Card"
        data-ouia-safe="true"
        id=""
        style="box-shadow: none;"
      >
        <div
          class="pf-c-card__body"
        >
          <div
            class="css-k008qs"
          >
            <div
              class="css-19gc7lg"
            >
              <svg
                aria-hidden="true"
                class="css-xeknh8"
                fill="var(--pf-global--warning-color--100)"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 576 512"
                width="1em"
              >
                <path
                  d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                />
              </svg>
            </div>
            <span
              style="margin-left: .4rem;"
            >
              <button
                aria-disabled="false"
                class="pf-c-button pf-m-link css-1dkl6e6"
                data-ouia-component-id="OUIA-Generated-Button-link-1"
                data-ouia-component-type="PF4/Button"
                data-ouia-safe="true"
                type="button"
              >
                Search is disabled on some clusters.
              </button>
            </span>
          </div>
        </div>
      </article>
    </div>
  </div>
</body>
`;

exports[`renders empty message 1`] = `
<body>
  <div>
    <div
      style="outline: none; display: flex; justify-content: flex-end;"
    >
      <div
        style="flex: 1;"
      >
        <section
          class="pf-c-page__main-section pf-m-no-padding pf-m-light"
        >
          <div
            class="pf-l-split"
          >
            <div
              class="pf-l-split__item pf-m-fill"
            >
              <div
                class="pf-l-stack pf-m-gutter"
              >
                <div
                  class="pf-l-stack__item pf-m-fill"
                >
                  <section
                    class="pf-c-page__main-section pf-m-light"
                  >
                    <div
                      class="pf-l-stack pf-m-gutter"
                    >
                      <div
                        class="pf-l-stack__item pf-m-fill"
                      >
                        <div
                          class="pf-l-split pf-m-gutter"
                        >
                          <div
                            class="pf-l-split__item"
                          >
                            <div
                              class="pf-c-content"
                            >
                              <h1
                                class="pf-c-title pf-m-2xl"
                                data-ouia-component-id="OUIA-Generated-Title-2"
                                data-ouia-component-type="PF4/Title"
                                data-ouia-safe="true"
                              >
                                Search
                              </h1>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </section>
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
  </div>
</body>
`;

exports[`renders unknown message 1`] = `
<body>
  <div>
    <div
      style="outline: none; display: flex; justify-content: flex-end;"
    >
      <div
        style="flex: 1;"
      >
        <section
          class="pf-c-page__main-section pf-m-no-padding pf-m-light"
        >
          <div
            class="pf-l-split"
          >
            <div
              class="pf-l-split__item pf-m-fill"
            >
              <div
                class="pf-l-stack pf-m-gutter"
              >
                <div
                  class="pf-l-stack__item pf-m-fill"
                >
                  <section
                    class="pf-c-page__main-section pf-m-light"
                  >
                    <div
                      class="pf-l-stack pf-m-gutter"
                    >
                      <div
                        class="pf-l-stack__item pf-m-fill"
                      >
                        <div
                          class="pf-l-split pf-m-gutter"
                        >
                          <div
                            class="pf-l-split__item"
                          >
                            <div
                              class="pf-c-content"
                            >
                              <h1
                                class="pf-c-title pf-m-2xl"
                                data-ouia-component-id="OUIA-Generated-Title-3"
                                data-ouia-component-type="PF4/Title"
                                data-ouia-safe="true"
                              >
                                Search
                              </h1>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </section>
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
      <article
        class="pf-c-card"
        data-ouia-component-id="OUIA-Generated-Card-2"
        data-ouia-component-type="PF4/Card"
        data-ouia-safe="true"
        id=""
        style="box-shadow: none;"
      >
        <div
          class="pf-c-card__body"
        >
          <div
            class="css-k008qs"
          >
            <div
              class="css-19gc7lg"
            >
              <svg
                aria-hidden="true"
                class="css-xeknh8"
                fill="var(--pf-global--warning-color--100)"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 576 512"
                width="1em"
              >
                <path
                  d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                />
              </svg>
            </div>
            <span
              style="margin-left: .4rem;"
            >
              <button
                aria-disabled="false"
                class="pf-c-button pf-m-link css-1dkl6e6"
                data-ouia-component-id="OUIA-Generated-Button-link-2"
                data-ouia-component-type="PF4/Button"
                data-ouia-safe="true"
                type="button"
              >
                Search is disabled on some clusters.
              </button>
            </span>
          </div>
        </div>
      </article>
    </div>
  </div>
</body>
`;

exports[`renders with Global Search alert & no message 1`] = `
<body>
  <div>
    <div
      style="outline: none; display: flex; justify-content: flex-end;"
    >
      <div
        style="flex: 1;"
      >
        <section
          class="pf-c-page__main-section pf-m-no-padding pf-m-light"
        >
          <div
            class="pf-l-split"
          >
            <div
              class="pf-l-split__item pf-m-fill"
            >
              <div
                class="pf-l-stack pf-m-gutter"
              >
                <div
                  class="pf-l-stack__item pf-m-fill"
                >
                  <section
                    class="pf-c-page__main-section pf-m-light"
                  >
                    <div
                      class="pf-l-stack pf-m-gutter"
                    >
                      <div
                        class="pf-l-stack__item pf-m-fill"
                      >
                        <div
                          class="pf-l-split pf-m-gutter"
                        >
                          <div
                            class="pf-l-split__item"
                          >
                            <div
                              class="pf-c-content"
                            >
                              <h1
                                class="pf-c-title pf-m-2xl"
                                data-ouia-component-id="OUIA-Generated-Title-4"
                                data-ouia-component-type="PF4/Title"
                                data-ouia-safe="true"
                              >
                                Global search
                                <button
                                  aria-disabled="false"
                                  class="pf-c-button pf-m-plain"
                                  data-ouia-component-id="OUIA-Generated-Button-plain-1"
                                  data-ouia-component-type="PF4/Button"
                                  data-ouia-safe="true"
                                  style="padding: 0px; margin-left: 8px; vertical-align: middle;"
                                  type="button"
                                >
                                  <svg
                                    aria-hidden="true"
                                    fill="currentColor"
                                    height="1em"
                                    role="img"
                                    style="vertical-align: -0.125em;"
                                    viewBox="0 0 512 512"
                                    width="1em"
                                  >
                                    <path
                                      d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                                    />
                                  </svg>
                                </button>
                              </h1>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </section>
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
  </div>
</body>
`;
